### ISSUE TITLE
Cannot assign to "None"

### ISSUE THREAD

**Issue by arulagrawal** [microsoft/pyright#1736]
**Describe the bug**
Pyright shows an error about not being able to assign a member of a class to an object of type class.

`Expression of type "Node" cannot be assigned to member "right" of class "Node"`
`Cannot assign to "None" (reportGeneralTypeIssues)`
**To Reproduce**
Run Pyright on the code below

**Expected behavior**
No errors, as python itself doesn't flag any errors and the program works.

**Screenshots or Code**
Code from https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/

https://gist.github.com/arulagrawal/6a41a8ebec16383215b822c692dfb4fc
**VS Code extension or command-line**
Command line pyright 1.1.128.

**Additional context**
This is a pretty basic example, so I assume I am doing something wrong here.

**Comment by erictraut** [comment]
If you do not provide a type annotation for an instance variable, pyright will infer its type from values assigned to it within the class definition. In this code, the `left` and `right` instance variables are assigned a value of `None`, so that is the inferred type. You can provide type annotations for variables where the type cannot be correctly inferred from assignments, like this:

```python
class Node:
    def __init__(self, key: int):
        self.left: Optional[Node] = None
        self.right: Optional[Node] = None
        self.val = key

def printPreorder(root: Optional[Node]): ...
```

If you are not interested in type checking but simply want completion suggestions and other language service features, you can set the `python.analysis.typeCheckingMode` setting to "off".
